### PR TITLE
MappingApplication new MapperFactory

### PR DESCRIPTION
**Pull request by philbucher**
@jcotela I break down the extensions in the MappingApp into several smaller PRs to make the review process easier

This PR contains mainly the new MapperFactory. The Interface changed, so I put the label API-Breaker (FYI @AndreasWinterstein @msandre @adityaghantasala). This should be the final API change, the rest is done internally.
Now the Mappers are used in python in the following way:
~~~py
import KratosMultiphysics.MappingApplication as KratosMapping

mapper = KratosMapping.MapperFactory(origin, destination, parameters) # old
mapper = KratosMapping.MapperFactory.CreateMapper(origin, destination, parameters) # new

mapper.Map(VAR1, VAR2, KratosMapping.MapperFactory.SWAP_SIGN) # old
mapper.Map(VAR1, VAR2, KratosMapping.Mapper.SWAP_SIGN) # new

~~~
The MapperFactory now is a real factory that returns a mapper. @pooyan-dadvand I remember that we were discussing abt the details of the implementation of this. I think you wanted some further changes, could you please take a look?

I did some reorganization, the content of the files `add_custom_utilities_to_python.*` was moved to `add_custom_mappers_to_python.*`, but stayed the same in principle (except the change that now the Mappers are exposed and not accessed through the factory any more)

Also some minor cleanup in the tests was done

@rubenzorrilla I already updated what I could find in the FSIapp. I didn't run the tests bcs I am not sure how you execute in mpi
Btw you can already now map on thin structrues, we did that several times already. You don't even need `POSITIVE_SURFACE_PRESSURE` and `NEGATIVE_SURFACE_PRESSURE`, you can directly map to `SURFACE_PRESSURE` with the following:

~~~py
# map first side
mapper1.Map(Var1, SURFACE_PRESSURE)
# map second side
mapper2.Map(Var1, SURFACE_PRESSURE, MappingApp.Mapper.ADD_VALUES) # this adds the values instead of overwriting them
# Repeat as often as needed
mapperN.Map(Var1, SURFACE_PRESSURE, MappingApp.Mapper.ADD_VALUES)
~~~

### CODE REVIEW

**Comment by loumalouomega** [comment]
@philbucher If this is the new interface, I will add a script to mimic this behaviour for the "simple_mortar__mapper", do you agree?

**Comment by philbucher** [comment]
@loumalouomega sounds good, how would you do it?

**Comment by loumalouomega** [comment]
The easiest way would be to create a python script with the same input despite it would work differently, and for the cpp it could be done similarly with an axillary constructor that replicates the behaviour of yours classes

**Comment by philbucher** [comment]
I think I have an idea, will make a draft later

**Comment by philbucher** [comment]
@loumalouomega we could make a mapper wrapper that wraps your process
This way the interface would be the same as for the mappers in the MappingApp
I have sth like this in mind:
~~~ cpp
class SimpleMortarMapperProcessWrapper : public Mapper

Map()
{
    // Call mapper of SimpleMortarMapperProcess
}

What do you think?
~~~

**Comment by loumalouomega** [comment]
For me is OK, but taking into account that is fully templatizated many explicit specialization will be needed

**Comment by rubenzorrilla** [comment]
@philbucher for me is OK (and thanks for updating the FSI app). Regarding the thin structures mapping, I will get rid of the positive and negative face pressure vars. in the way you point. Many thanks for the advise!

File: applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp
Comment (jcotela) [inline applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: The `inline` is fine (the compiler treats it only as a suggestion, anyway), but I'm not sure all those "dummy" functions are still needed: you are using the function pointers defined later to link to python.
If they are no longer required, please remove them.

**Review by jcotela — CHANGES_REQUESTED** [review]
Looks like a good clean-up to me 👍 
Can you just confirm if the dummy functions are still used and, if not, remove them?

File: applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp
Comment (philbucher) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: I need those functions to have the functions in Python exposed with the default arguments
e.g. the `UpdateInterface` Function I can call with 
~~~py
UpdateInterface()
UpdateInterface(Kratos::Flags)
UpdateInterface(double Search Radius)
UpdateInterface(Kratos::Flags, double Search Radius)
~~~
Same for the mapping functions

If you know an alternative, I am happy if I can achieve this behavior in an easier way

Comment (jcotela) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: I've never used it, but boost::python gives you support to do this automatically with `BOOST_PYTHON_MEMBER_FUNCTION_OVERLOADS`(see http://www.boost.org/doc/libs/1_65_1/libs/python/doc/html/tutorial/tutorial/functions.html#tutorial.functions.default_arguments.boost_python_member_function_ove).
If you want to keep doing it the way it is done now, it is OK for me too.

**Review by jcotela — DISMISSED** [review]
Approving, since I see you have a use for the dummy methods. Feel free to either merge as is or use the boost macro (and I'll re-approve).

File: applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp
Comment (philbucher) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: I can try to do it but I am not sure if it is worth the effort bcs I saw that @RiccardoRossi is starting to use pybind and when Kratos switches to pybind I could do the change properly
@RiccardoRossi what is the status of using pybind?

Comment (jcotela) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: @philbucher the pybind change won't be immediate, but I see your point. I hope it will have equivalent functionality however.
In any case, feel free to merge as is, it is OK for me.

Comment (philbucher) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: Hi @jcotela I tried the Boost Macro but it seems like it is not possible to use it. It seems to me like the macro creates some static object, which is not allowed in a local scope (i.e. `AddCustomMappersToPython`) in c++
~~~sh
/home/philippb/software/Kratos_aux/applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp:149:5: error: local class ‘struct Kratos::Python::AddCustomMappersToPython()::update_interface_overloads::non_void_return_type’ shall not have static data member ‘const int Kratos::Python::AddCustomMappersToPython()::update_interface_overloads::non_void_return_type::n_funcs’ [-fpermissive]
     BOOST_PYTHON_MEMBER_FUNCTION_OVERLOADS(update_interface_overloads, UpdateInterface, 0, 2);
~~~
I did a search, it does not seem to be used anywhere else in Kratos.
Have you used this somwhere else?

Comment (jcotela) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: No, I have never used it (I tend to not provide default arguments). It is just that I saw it in the boost docs as the "recommended" usage. If it doesn't work, just forget about it.

Comment (philbucher) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: Lets see maybe I find another way to do it better in the future

Then I am waiting for @pooyan-dadvand s approval

Comment (pooyan-dadvand) [reply applications/MappingApplication/custom_python/add_custom_mappers_to_python.cpp]: I have used this macro in some points when I started with python interface (don't ask me where) but afterward I arrived to the conclusion that the use of local functions is more clear and easier to debug. Personally I give them even different name (UpdateInterface1, UpdateInterface2, ...) to avoid the pointer function definition. (This was my personal flavor)

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
As you have asked my opinion about the factory I sould say that it needs modification. Actually the factory pattern is to avoid:
```c++
if("nearest_neighbour") 
    CreateNearestNeighbour(....)
```
which are not extensible and maintainable. The "by manual" way to do is to create a singleton list in your `MapperFactory` class which keeps the the name you are looking for and a creator function or prototype to be cloned:

```c++
class MapperFactory{
  static std::unordered_map<...>& GetRegisteredMapperList(){
    static std::unordered_map<...> s_registered_mapper;
    return s_registered_mapper;
  }
public:
  void Register(std::string Name, Mapper::Pointer Prototype) {
    GetRegisteredMapperList().insert(make_pair(Name, Prototype);
  }

  Mapper::Pointer Create(std::string const& Name, Parameters MapperParameters){
     if((auto i_mapper = GetRegisteredMapperList().find(Name)) != end)
         return i_mapper.Clone(MapperParameters);
   KRATOS_ERROR << Name << " Mapper not registered";
};
```

A far better explanation is given in Modern C++ Design by Andrei Alexandrescu

**Comment by philbucher** [comment]
@pooyan-dadvand I implemented your suggestion, can you please have a look?
Thx btw a lot for your suggestion!

File: applications/MappingApplication/mapping_application.h
Comment (pooyan-dadvand) [inline applications/MappingApplication/mapping_application.h]: This should be the `MapperFactory::Register` method and I would put it in the cpp of mapper factory. In future we will move this register to the core to enforce better the singleton and keep track of all objects.

File: applications/MappingApplication/mapping_application.h
Comment (pooyan-dadvand) [inline applications/MappingApplication/mapping_application.h]: This should be a private method of MapperFactory class

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]

**Comment by philbucher** [comment]
@pooyan-dadvand done, can you please have a look

**Review by pooyan-dadvand — COMMENTED** [review]
:+1: 

Just very minor thing. I would leave the name of the `RegisterMapper` method to `Register` so it would be the same as the future core registery mechanism

**Comment by philbucher** [comment]
@pooyan-dadvand done

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Pull request merged**
